### ISSUE TITLE
TLS Certificate sent to client is different from the one provided to server

### ISSUE THREAD

**Issue by siddjain** [osixia/docker-openldap#315]
we are seeing a strange issue where the TLS certificate sent to client is slightly different from the original certificate provided to the server. It results in certificate signature failure on the client. Wondering if anyone has encountered this and know how to fix it?

The difference is very slight - the order of OUs is swapped but nevertheless it will cause the signature to not match. we don't know if the problem is with openldap-docker or with openldap itself.

**Comment by obourdon** [comment]
@siddjain may be you could post the original certificate part with the OUs in non-swapped order, the part in swapped order which is sent to the client according to your statements above as well as the commands you are using to launch docker-openldap as well as the client command you use to connect. You can also try to use openssl command line utility to verify all this independently (openssl s_client -connect ...)

**Comment by siddjain** [comment]
@obourdon we have documented complete steps to repro the bug [here](https://github.com/siddjain/openldap-bug) with container logs. Thanks for any help.

**Comment by quanah** [comment]
This is most likely a bug in GnuTLS and would have to be kicked back upstream to them

**Comment by obourdon** [comment]
@quanah not sure what you really mean by this ? Any reference to existing issue that you might want to share ? Thanks

**Comment by quanah** [comment]
@obourdon IIRC, Osixia is using the OpenLDAP build from debian/ubuntu, which links to GnuTLS, rather than OpenSSL.  My point was, this would sound like an issue at that layer, rather than OpenLDAP itself.

**Comment by quanah** [comment]
@obourdon I discussed this with @rtandy who manages the OpenLDAP build for Debian, and he's fairly certain this is a bug in the version of GnuTLS in Debian 9 that's believed fixed for buster.

Generally the OpenLDAP project advises linking builds to OpenSSL.

**Comment by obourdon** [comment]
@quanah @rtandy thanks for the explanation which now makes a lot of sense to me. However, if I am not mistaken, I think I remember trying running OpenLDAP on debian-stretch in a VM (as opposed to a docker container) with the given certificates without any issue. But I am not 100% sure anymore.

**Comment by rtandy** [comment]
Hello @siddjain @obourdon, sorry for the delayed response.

First off, this is certainly not an OpenLDAP or docker-openldap issue, as it can be reproduced easily using only GnuTLS' own tools: gnutls-serv and gnutls-cli.

I tried a few different scenarios using the test certificates from https://github.com/siddjain/openldap-bug.

The connection always succeeds when I test it using openssl s_server/s_client:
```
openssl s_server -cert jnj-ldap-server-tls.pem -key jnj-ldap-server-tls.key -port 9999
openssl s_client -CAfile jnj-ca-chain.pem -connect localhost:9999
```

But if I mix in one or both of gnutls-serv/gnutls-cli, it always fails:
```
gnutls-serv --x509certfile jnj-ldap-server-tls.pem --x509keyfile jnj-ldap-server-tls.key -p 9999
gnutls-cli --x509cafile jnj-ca-chain.pem -p 9999 localhost
```

I see this on Debian 9 (stretch) and everything more recent, including the current devel branch. Sorry @quanah, I was wrong about it being fixed...

The problem is with the multi-valued OU attribute, "OU = client + OU = jnj". GnuTLS consistently renders it as "OU=jnj+OU=client". I guess it fails even when using both gnutls-serv and gnutls-cli because the signature on the certificate was computed with OpenSSL's ordering of the values, like @siddjain said.

There is a GnuTLS issue that looks very closely related: https://gitlab.com/gnutls/gnutls/issues/403. However this is claimed to be fixed in GnuTLS 3.6.3, whereas I still see the issue in 3.6.7 and later.

There is also a Stack Overflow question with some more context in the comments: https://stackoverflow.com/q/30193511.

Anyway, bottom line is this is neither an OpenLDAP issue nor a docker-openldap issue and therefore this ticket should be closed. @siddjain, if you still need this fixed, I recommend taking it up directly with the GnuTLS developers.
